### ISSUE TITLE
Plugin activation

### ISSUE THREAD

**Issue by JordanMarr** [MoiraeSoftware/myriad#98]
I think that the way I am activating my "ssdt" Myriad plugin for SqlHydra is a little different from your examples.
For example, you usually activate your plugin by marking an F# object with an attribute (i.e. `[<Generator.DuCases>]`).

But in the case of SqlHydra "ssdt" plugin, my source `<MyriadFile>` is a .dacpac file.  
So far this has worked fine because I only have one plugin in SqlHydra.dll. 
The issue is that, currently, my plugin will only work if it is the only plugin.

Last night I had some problems when I started creating a second plugin.
For starters, the new plugin doesn't even require an input file at all.  Really just needs to get some stuff out of the .toml config and then run on build. But there is no workflow currently to handle this kind of plugin since the `<MyriadFile>` usually points an fs file containing an attribute that will activate a named plugin.  In my scenario, I had two plugins, one that points to a .dacpac file, and another that doesn't need an input file at all, and Myriad seemed to be just using the first plugin it could find.
(I was able to comment one out which would then allow me to use the other). 

So I'm thinking that perhaps both my plugin workflows could be accommodated with an additions to the MSBuild configuration.

## Plugin Activation: non-fsharp MyriadFile
This style of plugin relies on a non-fsharp file (i.e. dacpac), and so it needs a way to specify the name of the plugin to run from the MSBuild config.  Maybe something like this:
```xml
    <ItemGroup>
         <!-- Specify the .fs output file (to be generated by Myriad) -->
        <Compile Include="AdventureWorks.fs">
            <!-- Specify the .dacpac input  file -->
            <MyriadFile>../AdventureWorks/bin/Debug/AdventureWorks.dacpac</MyriadFile>
            <MyriadPlugin>ssdt<MyriadPlugin>
        </Compile>
    </ItemGroup>
```

## Plugin Activation: no MyriadFile
This style of plugin can get all the input information it needs via a toml config section; it does not need an input `MyriadFile` at all.
Maybe this case would just need:
```xml
    <ItemGroup>
         <!-- Specify the .fs output file (to be generated by Myriad) -->
        <Compile Include="Generated.fs">
            <!-- No input file needed - just specify the plugin only -->
            <MyriadPlugin>newplugin<MyriadPlugin>
        </Compile>
    </ItemGroup>
```

**Comment by 7sharp9** [comment]
Currently the CLI input has mandatory input so that would have to be addressed first:  https://github.com/MoiraeSoftware/myriad/blob/master/src/Myriad/Program.fs#L14

Actual the MSBuild target property is also mandatory too.

plugins are included via the msbuild injection of props from the nuget `generator.props` etc

**Comment by JordanMarr** [comment]
Interesting.. So now I see how you are binding the Plugin arg from MSBuild to CLI [here](https://github.com/MoiraeSoftware/myriad/blob/master/src/Myriad.Sdk/build/Myriad.Sdk.targets):
```xml
<MyriadSdk_Args Include='--plugin %(MyriadSdkGenerator.FullPath)' Condition=" '@(MyriadSdkGenerator)' != '' "/>
```

That is passing the plugin .dll path to the Myriad CLI -- which I am doing that here in [SqlHydra.props](https://github.com/JordanMarr/SqlHydra/blob/main/src/SqlHydra/build/SqlHydra.props).

But I have multiple plugins in my dll, so then how can it know which to use? 
Is MSBuild going to try to execute all of the plugins it finds in my dll?
Can more than one `ItemGroup` be configured within a single project? (like if I wanted to use two different plugins in the same prjoect)?

**Comment by 7sharp9** [comment]
All plugins in a dll will get called, fields/dus/lend all share the same dll

**Comment by JordanMarr** [comment]
Gotcha. Sorry, I was referring to generators as plugins which was adding to the confusion.
All generators get called on build.

**Comment by JordanMarr** [comment]
So then let me try again:
Would it make sense to be able to specify a `MyriadGenerator` since a plugin dll can have multiple generators?

```xml
    <ItemGroup>
         <!-- Specify the .fs output file (to be generated by Myriad) -->
        <Compile Include="AdventureWorks.fs">
            <!-- Specify the .dacpac input  file -->
            <MyriadFile>../AdventureWorks/bin/Debug/AdventureWorks.dacpac</MyriadFile>
            <MyriadGenerator>ssdt<MyriadPlugin>
        </Compile>
    </ItemGroup>
```

**Comment by 7sharp9** [comment]
>But I have multiple generators in my plugin dll, so then how can it know which to use?

Thats a good question when theres no input file, the addition of validExtension was added as part of that filter for input files that didn't have a way of attributing what needed generating, I think input file could be empty is the CLI argument was non mandatory and the msbuild condition was removed.  I think some sort of modulation of the props might be needed, I did once thing that having key/values in the msbuild could be useful but hoped the toml config might be better.

```xml
Compile.MyriadFile
```

```xml
Compile.Myriad.File
```

Having Myriad as the basic element might make more sense now that fs/any/none could be valid input types

**Comment by 7sharp9** [comment]
What I wanted to make sure of was that complexity didn't explode, once you add two ways of doing things the logic starts to twist, we would have 3 ways of invoking a plugin, and three ways of specifying sun properties which all have to end up being CLI params.

**Comment by 7sharp9** [comment]
```xml
    <ItemGroup>
         <!-- Specify the .fs output file (to be generated by Myriad) -->
        <Compile Include="AdventureWorks.fs">
            <!-- Specify the .dacpac input  file -->
            <Myriad>
                <InputFile>../AdventureWorks/bin/Debug/AdventureWorks.dacpac</MyriadFile>
                <Generator>
                    <Name>ssdt</Name>
                    <Namespace>test</Namespace>
                </Generator>
                <!-- Or just text element -->
                <Generator>other</Generator>
            <Myriad>
        </Compile>
    </ItemGroup>```
    
Theres also using the toml file to add all these extra pieces of info:
    

<ItemGroup>
     <!-- Specify the .fs output file (to be generated by Myriad) -->
    <Compile Include="AdventureWorks.fs">
        <!-- Specify the .dacpac input  file -->
        <Myriad>
            <Generator>
                <Config>myGeneratorSection</Config>
            </Generator>
        <Myriad>
    </Compile>
</ItemGroup>
```

**Comment by 7sharp9** [comment]
I think maybe extending the interface so that IsValid is passed some properties and then the plugin author can decide whether its applied of not

`IsValid(extension: string option, ...) : bool`

**Comment by 7sharp9** [comment]
It could also be possible to specify a glob as part of an filename to tell the generator what to do like `[id].fs` which would give it a one to many output.  Im thinking too far ahead anyway, need to reign it in :-)

**Comment by JordanMarr** [comment]
The clarity of intent is really high in this example you posted above. I like it!

```xml
    <ItemGroup>
        <Compile Include="AdventureWorks.fs">
            <Myriad>
                <InputFile>../AdventureWorks/bin/Debug/AdventureWorks.dacpac</InputFile>
                <Generator>ssdt</Generator>
            <Myriad>
        </Compile>
    </ItemGroup>
```

Things I like about it:
- It doesn't need my comments anymore because it's very obvious that Myriad will be generating the file, and it's obvious that all the elements within `<Myriad>` are configuration settings.
- `<InputFile>` has more clear intent than `<MyriadFile>` (which isn't immediately obvious whether it is for input or output). This is less an issue in my scenario since the input is a .dacpac, but worse if both input and output are .fs files.
- It would allow specifying a named generator in the case where no input file is needed
- Moving Namespace configuration into here seems pretty convenient. Although, the toml is already pretty straight forward and flexible.

**Comment by JordanMarr** [comment]
> I think maybe extending the interface so that IsValid is passed some properties and then the plugin author can decide whether its applied of not
> 
> `IsValid(inputFile: string option, extension: string option, generators: string list) : bool`
> 
> `generators` would probably be exclusive if **present** I guess its up to the plugin to decide based on the inputs passed

The `IsValid` filter is also a very nice idea. It seems like maybe it could even replace the `ValidInputExtensions` member altogether since it has the `extension` passed to it.
I like that `inputFile` is optional. I'm not sure how `generators` would work so I have opinion on it.

**Comment by 7sharp9** [comment]
><InputFile> has more clear intent than <MyriadFile> (which isn't immediately obvious whether it is for input or output). This is less an issue in my scenario since the input is a .dacpac, but worse if both input and output are .fs files.

Its to do with how msbuild work, output file is Compile Include, it has to be that way to get any integration

Technically it could just be `<Input>{ ... }</Input>` with inline data which is output, I think the file suffix is a bit wordy.  Having `Myriad` as the parent makes it obvious that the siblings are just for it.

**Comment by 7sharp9** [comment]
I think making the input optional and adding explicit Generators after this:
https://github.com/MoiraeSoftware/myriad/blob/master/src/Myriad.Sdk/build/Myriad.Sdk.targets#L8

Is probably the way to proceed.

**Comment by 7sharp9** [comment]
At some point I'll draft a spec, or someone else could if they feel inclined, and start a PR.  Im not sure when I'll have time just yet.

**Comment by 7sharp9** [comment]
@JordanMarr Im not entirely sure what the conclusion of all this was?

**Comment by 7sharp9** [comment]
Maybe its:  There needs to be a way of specifying which generators do or don't run on arbitrary files, so some type of filter, and as it cant be in the file itself it needs to be in the fsproj, hence the need for a generator and maybe an exludeGenerator element?

```
    <Compile Include="ArbitaryFile.fs">
      <MyriadFile>Test.txt</MyriadFile>
      <MyriadConfigKey>example1</MyriadConfigKey>
      <MyriadParams>
        <MyriadParam1>5</MyriadParam1>
        <MyriadParam2>6</MyriadParam2>
      </MyriadParams>
    </Compile>
 ```
 Thats what an arbitrary  input file currently supports

**Comment by 7sharp9** [comment]
It does feel like a feature hole where everything just runs, especially it you like to cram multiple plugins in one dll.

**Comment by 7sharp9** [comment]
Thinking more I don't think its even possible at the MSBuild lever as Myriad uses reflection to find all plugins inside the passed dll, so it has to be done at that point which is annoying.

**Comment by 7sharp9** [comment]
This is done, just need docs 😢

**Comment by JordanMarr** [comment]
Or you could just throw a few lines into the readme and be done with it.

**Comment by 7sharp9** [comment]
CAnt remember anything about this, so closing...
